### PR TITLE
support contraction of arbitrary objects

### LINKED ISSUE #86: Potential bug (outer tensor product)
This works...:

```python
>>> import mpmath
>>> import numpy
>>> m0 = numpy.eye(2)
>>> opt_einsum.contract('bc,ad->abcd', m0, m0)
array([[[[1., 0.], ...)
```

...but this breaks:

```python
>>> m1 = m0 + mpmath.mpc(0)
>>> opt_einsum.contract('bc,ad->abcd', m1, m1)
Traceback (most recent call last): (...)
TypeError: invalid data type for einsum
```

However, using opt_einsum with mpmath is otherwise OK in principle:

```python
>>> opt_einsum.contract('bc,cd->bd', m1, m1)
array([[mpc(real='1.0', imag='0.0'), mpc(real='0.0', imag='0.0')],
       [mpc(real='0.0', imag='0.0'), mpc(real='1.0', imag='0.0')]],
      dtype=object)
```

### PR DESCRIPTION
**Pull request by jcmgray**
## Description
Add ``backend='object'`` to support contraction of arrays of arbitrary python objects (resolves #86 and #142).

## Status
- [x] Ready to go

### CODE REVIEW

File: docs/source/backends.rst [hunk 1]
```diff
@@ -325,3 +325,47 @@ so again just the ``backend`` keyword needs to be supplied:
...
+  >>> X = np.array([[a, b], [c, d]])
+  >>> Y = np.array([[e, f], [g, h]])
+  >>> Z = np.array([[i, j], [k, l]])
+
+  >>> # contract the tensors!
+  >>> oe.contract('uv,vw,wu->u', X, Y, Z, backend='object')
+  array([i*(a*e + b*g) + k*(a*f + b*h), j*(c*e + d*g) + l*(c*f + d*h)],
+        dtype=object)
+
+There's a few things to note here:
```
Comment (dgasmith) [inline docs/source/backends.rst:362]: "There are"

File: docs/source/backends.rst [hunk 2]
```diff
@@ -325,3 +325,47 @@ so again just the ``backend`` keyword needs to be supplied:
...
+
+Contracting arbitrary objects
+=============================
+
+There is one more explicit backend that can handle arbitrary arrays of objects,
+so long the *objects themselves* just support multiplication and addition (
+``__mul__`` and ``__add__`` dunder methods respectively). Use it by supplying
+``backend='object'``.
+
+For example imagine we want to perform a contraction of arrays made up of
```
Comment (dgasmith) [inline docs/source/backends.rst:338]: "For example,"

File: opt_einsum/backends/objects.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+"""
+Functions for performing contractions with array elements which are objects.
+"""
+
+import numpy as np
+import functools
+import operator
+
+
+def oinsum(eq, *arrays):
```
Comment (dgasmith) [inline opt_einsum/backends/objects.py:10]: `object_einsum`? Something slightly more explicit IMO, ditto for the filename.

File: opt_einsum/backends/objects.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    """A ``einsum`` implementation for ``numpy`` arrays with object dtype.
+    The loop is performed in python, meaning that the objects themselves need
```
Comment (dgasmith) [inline opt_einsum/backends/objects.py:12]: "meaning the objects"

File: opt_einsum/tests/test_backends.py [hunk 5]
```diff
@@ -431,3 +431,25 @@ def test_auto_backend_custom_array_no_tensordot():
     # Shaped is an array-like object defined by opt_einsum - which has no TDOT
     assert infer_backend(x) == 'opt_einsum'
     assert parse_backend([x], 'auto') == 'numpy'
+
+
+@pytest.mark.parametrize("string", tests)
+def test_objects(string):
```
Comment (dgasmith) [inline opt_einsum/tests/test_backends.py:437]: `test_object_backend`

File: opt_einsum/backends/objects.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        These can be any indexable arrays as long as addition and
+        multiplication is defined on the elements.
+
+    Returns
+    -------
+    out : numpy.ndarray
+        The output tensor, with ``dtype=object``.
+    """
+
+    # when called by ``opt_einsum`` we will always be given a full eq
```
Comment (dgasmith) [inline opt_einsum/backends/objects.py:32]: This is pretty impressively short to pass *all* of our tests which have some odd operations in them.

**Comment by jcmgray** [comment]
Thanks for the comments, should have addressed them all in latest commit.

**Pull request merged**